Project configuration input files

aspeake opened the pull request:
Update how arguments for ecm_prep.py and run.py are handled. A configuration .yml file is introduced to simplify running these scripts, improve repeatability, and allow for version control. New arguments are introduced as needed to replace the user prompts and responses. Additionally, a new set of tests is added to cover these new features and as well as gaps in testing for existing argument handling.

**Summary of Changes**
- New scout/config.py to: 
  - digest yml inputs files and maps to Scout arguments
  - validate the yml against a custom schema
  - translate the schema file to be read with the --help flag
- Update scout/ecm_prep_args.py to remove all user prompts and utilize config.py
- New tests/test_config.py to test new features in config.py and test existing features for handling args in ecm_prep.py and run.py

**Functionality Changes**
- Arguments for ecm_prep.py and run.py can be defined with either a .yml file (e.g., `python scout/ecm_prep.py -y <myyml>.yml`) or by passing CLI arguments directly
- User prompts and responses are no longer part of the argument handling in ecm_prep.py and run.py

**Example yml Input:**
```
description: This is an example config file to run ecm_prep.py and/or run.py
ecm_prep:
  site_energy: true
  add_typ_eff: true
  rp_persist: true
  alt_regions: EMM
  adopt_scn_restrict: Max adoption potential
  fugitive_emissions: [methane, typical refrigerant]
  retrofits:
    retrofit_type: increasing
    retrofit_multiplier: 1.2
    retrofit_mult_year: 2030
  exog_hp_rates: 
    exog_hp_rate_scenario: aggressive
    switch_all_retrofit_hp: False
  grid_decarb:
    grid_decarb_level: full
    grid_assesment_timing: after
  tsv_metrics:
    tsv_type: energy
    tsv_daily_hr_restrict: all
    tsv_sys_shape_case: total reference
    tsv_season: summer
    tsv_energy_agg: average
    tsv_average_days: weekdays
  detail_brkout: [regions, fuel types]
  pkg_env_costs: include HVAC  

run:
  mkt_fracs: true
```

aspeake commented on scout/ecm_prep_args.py:
```diff
@@ -330,151 +130,48 @@ def fill_user_inputs(opts: argparse.NameSpace) -> argparse.NameSpace:  # noqa: F
+            ("regions",): "2",
+            ("buildings",): "3",
+            ("fuel types",): "4",
+            ("buildings", "fuel types", "regions"): "1",
```
change to "5"

aspeake commented on scout/ecm_prep_args.py:
```diff
@@ -330,151 +130,48 @@ def fill_user_inputs(opts: argparse.NameSpace) -> argparse.NameSpace:  # noqa: F
+    # Set detailed breakout options
+    if opts.detail_brkout:
+        input_var_dict = {
+            ("buildings", "regions"): "5",
```
change to "1"

aspeake commented on scout/ecm_prep_args.py:
```diff
@@ -330,151 +130,48 @@ def fill_user_inputs(opts: argparse.NameSpace) -> argparse.NameSpace:  # noqa: F
-            opts.detail_brkout = '3'
+    # Set detailed breakout options
+    if opts.detail_brkout:
+        input_var_dict = {
```
Confirmed that the numbers map appropriately to the correct use in ecm_prep.py

trynthink commented on scout/config.py:
```diff
@@ -0,0 +1,150 @@
+        parser (argparse.ArgumentParser): Argument parser
+        schema_file (Pathlib.filepath): Location of the configuration schema yml
+        schema_data (dict): Parsed schema data
+        key (str): Workflow step on which to validate and set arguments {ecm_prep, run}
```
Should there be a definition for the `cli_args` attribute?

aspeake replied:
Yes, updated.

trynthink commented on scout/config.py:
```diff
@@ -0,0 +1,150 @@
+            raise ValueError(
+                "Detailed breakout (detail_brkout) cannot include `fuel types` if split_fuel==True")
+
+    def create_argparse(self, parser, schema_data, group=None):
```
Is this function to enable backwards compatibility with the command line arguments approach previously in place? If so, can you add that detail to the docstring?

aspeake replied:
It is to automatically configure ArgumentParser with the schema, allowing --help to be populated and running with both yml and cli args. I've updated the docstring.

trynthink commented on scout/config.py:
```diff
@@ -0,0 +1,150 @@
+        """Extracts arguments from the config schema and writes argparse arguments"""
+        if group:
+            parser = group
+        arguments = copy.deepcopy(schema_data).get("properties", {})
```
Could you add a handful of comments to this function to help explain what the different parts are doing?

trynthink commented on scout/config.py:
```diff
@@ -0,0 +1,150 @@
+            config_key_args = self.config_args.get(self.key, {})
+            self.args = self.update_args(self.args, config_key_args)
+
+        # Ensure command-line args take precendence
```
"precendence" -> "precedence"

trynthink commented on scout/supporting_data/config_schema.yml:
```diff
@@ -0,0 +1,243 @@
+      floor_start:
+        type: ["integer", "null"]
+        default: null
+        description: The starting floor.
```
Change the description to something like: "Apply an elevated performance floor starting in the year specified. The desired future year must be given in the format YYYY."

trynthink commented on scout/supporting_data/config_schema.yml:
```diff
@@ -0,0 +1,243 @@
+      no_scnd_lgt:
+        type: boolean
+        default: false
+        description: Disable secondary lighting.
```
For booleans, how is the description phrased relative to the default? In this case, if the default is `false` and the description starts with "Disable" does that mean that setting to `true` "enables" the option in question?

trynthink replied:
Also rephrase this description to something like "(Enable/disable) the calculation of secondary heating and cooling energy effects from changes in lighting efficacy."

trynthink commented on scout/supporting_data/config_schema.yml:
```diff
@@ -0,0 +1,243 @@
+        default: false
+        description: Enable package environmental separation.
+
+      detail_brkout:
```
Can you put this right before/after "split_fuel" since they are related?

trynthink commented on scout/supporting_data/config_schema.yml:
```diff
@@ -0,0 +1,243 @@
+      split_fuel:
+        type: boolean
+        default: false
+        description: Enable split fuel.
```
Rephrase this description to something like: "(Enable/disable) ECM results reporting split out by fuel type."

trynthink commented on scout/supporting_data/config_schema.yml:
```diff
@@ -0,0 +1,243 @@
+      site_energy:
+        type: boolean
+        default: false
+        description: Enable site energy calculation.
```
For booleans, how is the description phrased relative to the default? If the default is `false` and the description starts with "Enable" does that mean that setting to `true` "disables" the option in question?

aspeake replied:
I updated all bools to start with "If true, ..." to improve clarity

trynthink commented on scout/supporting_data/config_schema.yml:
```diff
@@ -0,0 +1,243 @@
+      sect_shapes:
+        type: boolean
+        default: false
+        description: Enable sector shapes.
```
Rephrase to something like "Enable calculation of sector-level electricity savings shapes."

trynthink commented on scout/supporting_data/config_schema.yml:
The comments on the descriptions in `docs/config_readable.yml` should match up with the comments on this file.

trynthink commented on scout/supporting_data/config_schema.yml:
```diff
@@ -0,0 +1,243 @@
+          retrofit_type:
+            type: ["string", "null"]
+            enum: [constant, increasing, null]
+            description: Type of retrofit. `constant` assumes component-based early retrofit rates that do not change over time, `increasing` assumes rates that increase over time.
```
Rephrase "Type of retrofit." to something like "Specify early retrofit rates."

trynthink commented on scout/supporting_data/config_schema.yml:
```diff
@@ -0,0 +1,243 @@
+            enum: [constant, increasing, null]
+            description: Type of retrofit. `constant` assumes component-based early retrofit rates that do not change over time, `increasing` assumes rates that increase over time.
+          retrofit_multiplier:
+            type: ["number", "null"]
```
Is "number" a type? I would expect the non-integer numeric type to be "float."

aspeake replied:
Number is basically a float: https://json-schema.org/understanding-json-schema/reference/numeric#number

trynthink commented on scout/supporting_data/config_schema.yml:
```diff
@@ -0,0 +1,243 @@
+          grid_decarb_level:
+            type: ["string", "null"]
+            enum: ["0.8", full, null]
+            description: Enable grid decarbonization - `0.8` represents 80%% reduced grid emissions from current levels by 2050, and `full` represents full grid decarbonization by 2035.
```
Check that `%%` is not a typo.

trynthink commented on scout/supporting_data/config_schema.yml:
```diff
@@ -0,0 +1,243 @@
+      adopt_scn_restrict:
+        type: ["string", "null"]
+        enum: [Max adoption potential, Technical potential, null]
+        description: Specify adoption scenario restrictions.
```
Rephrase to "Specify single desired adoption scenario."

trynthink commented on scout/supporting_data/config_schema.yml:
```diff
@@ -0,0 +1,243 @@
+      add_typ_eff:
+        type: boolean
+        default: false
+        description: Enable additional typical efficiency.
```
Rephrase to something like "Enable automatic addition of ECMs with typical (business-as-usual) efficiency to compete with user-defined ECMs."

trynthink commented on scout/supporting_data/config_schema.yml:
```diff
@@ -0,0 +1,243 @@
+      pkg_env_sep:
+        type: boolean
+        default: false
+        description: Enable package environmental separation.
```
Rephrase to something like "Enable output of separate envelope ECM impacts when both HVAC and envelope ECMs are combined in one or more measure packages."

aspeake commented on docs/config_readable.yml:
```diff
@@ -0,0 +1,111 @@
+      to heat pumps, otherwise, retrofits are subject to the same
+      external heat pump conversion rates assumed for new/replacement
+      decisions. Only applicable if `exog_hp_rate_scenario` and
+      `retrofit_type` are defined. Default None
```
Good catch, all bool args will have an explicit default value now

aspeake commented on docs/config_readable.yml:
```diff
@@ -0,0 +1,111 @@
+      external heat pump conversion rates assumed for new/replacement
+      decisions. Only applicable if `exog_hp_rate_scenario` and
+      `retrofit_type` are defined. Default None
+  floor_start: (integer) The starting floor. Default 0
```
Using your suggested description for floor start in `config_schema.yml` (those descriptions match exactly to `config_readable.yml`)

aspeake commented on docs/config_readable.yml:
```diff
@@ -0,0 +1,111 @@
+      Default None
+    grid_decarb_level: (string, required if the grid_decarb key
+      is present) Enable grid decarbonization - `0.8` represents
+      80%% reduced grid emissions from current levels by 2050,
```
Was required for formatting in --help output, but doesn't work in translation to yml. Updating to "percent"

aspeake commented on tests/test_files/default_config.yml:
For our use, no it does not, but I updated for consistency

trynthink commented on docs/tutorials.rst:
```diff
@@ -1458,22 +1485,33 @@ Isolate W/E impacts in ECM packages
 
+Alternative Emissions Intensities
+*********************************
+``--alt_ref_carb`` sets the baseline electricity emissions intensities to be consistent with the Standard Scenarios Mid Case (rather than AEO).
```
The 2023 Standard Scenarios report describes the mid case. https://www.nrel.gov/docs/fy24osti/87724.pdf

aspeake commented on docs/config_readable.yml:
```diff
@@ -0,0 +1,111 @@
+    retrofit_mult_year: (integer) The year by which the retorfit
+      multiplier is achieved (only for increasing retrofit_type).
+      Default None
+    retrofit_multiplier: (number) Factor by which early retrofit
```
"number" is analogous to a python float in jsonschema

aspeake commented on docs/config_readable.yml:
For "adopt_scn_restrict" the description has been updated to clarify. "alt_regions" will be updated in a future PR as discussed: https://github.com/trynthink/scout/issues/275#issuecomment-1909044835

aspeake commented on docs/config_readable.yml:
```diff
@@ -0,0 +1,111 @@
+      `retrofit_type` are defined. Default None
+  floor_start: (integer) The starting floor. Default 0
+  fugitive_emissions: (array) Array enabling fugitive emissions;
+    array may include methane and one of the two refrigerant options.
```
This was a missed case for arrays in generate_readable_cfg.py, and has been corrected.

aspeake commented on docs/generate_readable_cfg.py:
```diff
@@ -0,0 +1,42 @@
+output_file = Path(__file__).resolve().parents[0] / "config_readable.yml"
+
+
+def convert_yaml_structure(input_yaml, required=None, parent_key=None):
```
Added better docstring and some comments

aspeake commented on tests/arguments_test.py:
Added more descriptive docstrings and comments for certain tests

aspeake commented on scout/config.py:
Agreed. I added better docstrings and type hinting.

aspeake commented:
@trynthink thanks for the review. I think I've addressed each of your comments and do not have any followups. I added a CI job to check that the readable config in the docs has been updated for any changes with the schema: https://github.com/trynthink/scout/pull/355/commits/4d89032b0fcc6820d91a1ce7a4d66d2c5bbd3648. Given I didn't even update it in this PR, it seems like an easy thing to miss. I stopped short of an automatic commit to update, and instead just fail the CI.

aspeake commented on .github/workflows/tests.yml:
```diff
@@ -40,21 +40,46 @@ jobs:
+      - name: Get contents from tip of branch on push
+        uses: actions/checkout@v4
+        with:
+          sparse-checkout: |
+            scout
+            docs
```
fyi - easy way to avoid checking out the full repo w/ ecm_definitions if they are not needed.

trynthink commented on docs/installation_guide.rst:
@aspeake We should amplify this guidance to the rest of the team either at our next check-in or on chat.

aspeake replied:
Agreed, I plan on posting on Slack once merged.

trynthink commented on scout/config.py:
```diff
@@ -0,0 +1,217 @@
+                        schema_data: dict, group: str = None):
+        """Extracts arguments from the config schema and writes argparse arguments. This method
+            populates information for the --help flag for ecm_prep.py and run.py and enables
+            passing arguments directly to the command line.
```
"directly to" or "directly at" the command line?

trynthink commented on scout/config.py:
```diff
@@ -0,0 +1,217 @@
+        return existing_args
+
+    def check_dependencies(self, args: argparse.NameSpace):  # noqa: F821
+        """Allow for custom checks that are not easily defined directly in the schema
```
Can you add another sentence or two explaining why these custom checks have to be done here? (I'm aiming for some additional information so that another developer who sees this function will know when to add something here.)

aspeake commented on scout/ecm_prep_args.py:
```diff
@@ -174,109 +62,44 @@ def fill_user_inputs(opts: argparse.NameSpace) -> argparse.NameSpace:  # noqa: F
-                if input_var[1] not in ['1', '2']:
-                    print('Please try again. Enter either 1 or 2. '
-                          'Use ctrl-c to exit.')
+        for emission in opts.fugitive_emissions:
```
Corrected in https://github.com/trynthink/scout/pull/355/commits/edf04f50593bfa7eb63981911a58e9671f6d97d8

aspeake commented on docs/config_readable.yml:
Corrected in https://github.com/trynthink/scout/pull/355/commits/c273244884b17d151e0e9a19e3f7a60e2b06d811

trynthink commented on tests/arguments_test.py:
```diff
@@ -0,0 +1,395 @@
+        self._assert_arg_vals(args, self.valid_yml_translated)
+
+    def test_translate_cli_overwrite(self):
+        # Ensure cli arguments take precedence over yml
```
```suggestion
        # Ensure cli arguments take precedence over yml and are correctly translated
```

You could use different words here, but my initial reaction was "how is this test different from `TestConfig.test_cli_overwrite()`" so the change is proposed to hopefully prevent that misinterpretation.

aspeake commented on tests/arguments_test.py:
```diff
@@ -0,0 +1,395 @@
+        expected_err = "'description' is a required property"
+        self._check_schema_err(expected_err, drop_key="description")
+
+    def test_invalid_config_args(self):
```
Note - I added a few more asserts in this test after the dependency check was expanded (https://github.com/trynthink/scout/pull/355/commits/50668a94801401959d418af4aa2e6be6a0df1ede). A new method in arguments_test, `_get_cfg_args_err_message()`, was added to facilitate this.
